Zoom Layout Enhancements (GitHub issue #2982)

geowar1 opened the pull request:
Note: I added a test case and schema for LayoutEditor:setZoom(); As a noob JMRI developer I'd appreciate a (peer) code review on that (if possible). Thanks.

geowar1 commented:
Note: I added a test case and schema for LayoutEditor:setZoom(); As a noob JMRI developer I'd appreciate a (peer) code review on that. Thanks.

geowar1 commented:
Note: I added a test case and schema for LayoutEditor:setZoom(); As a noob JMRI developer I'd appreciate a (peer) code review on that (if possible). Thanks.

geowar1 commented:
The Travis CI build failed with     [junit] ERROR - run: Exception: java.io.IOException: Pipe broken in class jmri.jmrix.roco.z21.Z21XNetPacketizer [testShowAndClose] jmri.jmrix.AbstractMRTrafficController.?()
That's nothing that I edited… so… Try, try again?

rhwood commented:
This make the zoom a direct property of the panel. Is that what was intended? You most likely want to store the zoom in the UserPreferencesManager as a property of the window containing the panel instead--this will allow that property to be per computer.

rhwood commented:
You also want to remove the verticalToolbar setting from the GuiLafPreferencesManager and put it in the UserPreferencesManager--doing so would allow the verticalToolbar setting to be set per window, instead of forcing it on all windows.

geowar1 commented:
Ready to merge… TIA.

geowar1 commented:
Maybe not (ready to merge); I just saw rhwood's comments… let me look into them before we close this issue.

geowar1 commented:
rhwood: You're going to have to help me out a little here… I'm too much a noob to write virgin code and I'm not seeing the UserPreferencesManager used anywhere in LayoutEditor.java to plagiarize. LayoutEditor extends PanelEditor extends Editor extends JmriJFrame which uses the UserPreferencesManager in the setFrameLocation routine... looks like the windowFrameRef is used to make a setting per-window? So would the code be:
`

        // get the window specific saved zoom user preference

        InstanceManager.getOptionalDefault(UserPreferencesManager.class).ifPresent(prefsMgr -> {

            String windowFrameRef = getWindowFrameRef();

            log.info("windowFrameRef: " + windowFrameRef);

            String prefsZoomString = prefsMgr.getProperty(windowFrameRef, "zoom").toString();

            if (prefsZoomString != null) {

                double prefsZoom = Double.parseDouble(prefsZoomString); 

                setZoom(prefsZoom);

            }

        });
`
Hummm… the prefsMgr.getProperty call is throwing a java.lang.NullPointerException… and I don't know how to fix it… Little help?

rhwood commented:
Try something more like:

```
    // get the window specific saved zoom user preference
    InstanceManager.getOptionalDefault(UserPreferencesManager.class).ifPresent(prefsMgr -> {
        String windowFrameRef = getWindowFrameRef();
        log.debug("windowFrameRef: " + windowFrameRef);
        Object zoomProp = prefsMgr.getProperty(windowFrameRef, "zoom");
        if (zoomProp != null) {
            // need try/failure for number format exception if zoom property is not a number
            double zoomPref = Double.parseDouble(zoomProp.toString());
            if (zoomPref > minZoom && zoomPref < maxZoom) {
                this.setZoom(zoomPref);
            }
        }
    });
```

I think the NPE was caused by calling toString() on the null result of the prefsMgr.getProperty() call, not from within the prefsMgr.getProperty() call itself.

geowar1 commented:
On Feb 10, 2017, at 10:57 AM, Randall Wood <notifications@github.com> wrote:
> …
> I think the NPE was caused by calling toString() on the null result of the prefsMgr.getProperty() call, not from within the prefsMgr.getProperty() call itself.

Gerrr… I should have figured that out; I was assuming it was the getProperty call that was throwing the exception…

Now getProperty is always returning null…

Is my save code correct? (Note: I changed the zoomProp from a String to a Double (to avoid the parsing number format exception).)

            // save the window specific saved zoom user preference
            InstanceManager.getOptionalDefault(UserPreferencesManager.class).ifPresent(prefsMgr -> {
                prefsMgr.setProperty(getWindowFrameRef(), "zoom", Double.valueOf(zoomFactor));
            });

This is my get code now:

        // get the window specific saved zoom user preference
        InstanceManager.getOptionalDefault(UserPreferencesManager.class).ifPresent(prefsMgr -> {
            Object zoomProp = prefsMgr.getProperty(getWindowFrameRef(), "zoom");
            if (zoomProp != null) {
                this.setZoom(((Double) zoomProp).doubleValue());
            }
        });

(Note: I moved the "if (zoomFactor >= minZoom && zoomFactor <= maxZoom) {“ into the setZoom routine.)

-- 
Enjoy,
George Warner, (408)249-7533
eMail/iChat: <geowar1@mac.com>

geowar1 commented:
Unless rhwood objects… I think I'm going to close this issue and open a new one to migrate some settings to use UserPreferencesManager.

KenC57 commented:
I wouldn't say closing the issue, just the PR that was going to do it via
edits to the panel instead of things with the user preferences. It is really
the same issue at the heart, just a different route to solve it.

-ken c

rhwood commented:
LGTM
